src/Controllers/cli_controller.py

```diff
@@ -26,11 +30,161 @@ def parse_list_cmd(self, d:UML_Diagram, tokens:list[str]):
             case _:
                 raise ValueError("Invalid use of list.")
 
-    def parse_help_cmd(self, tokens:list[str]):
+    def parse_help_cmd(self, tokens: list[str]):
         """Parses a help command, returning it in a form ready to be run"""
         if len(tokens) == 0:
             return [getattr(self._view, 'help')]
         elif len(tokens) == 1:
             return [getattr(self._view, 'help_' + tokens.pop(0))]
         else:
-            raise ValueError("Invalid command.")
\ No newline at end of file
+            raise ValueError("Invalid command.")
+
+
+# ========================= Autocomplete Functionality =========================#
+
+
+    def setup_autocomplete(self):
+        """
+        Sets up autocomplete functionality using the NestedCompleter.
+
+        """
+
+        self._completer = NestedCompleter.from_nested_dict({
+            'add': {
+                'relation': self.get_add_relation(),
+                'class': None,
+                'field': self.get_classes(),
+                'method': self.get_classes(),
+                'param': self.get_methods()
+
+            },
+            'delete': {
+                'relation': self.get_delete_relation(),
+                'class': self.get_classes(),
+                'field': self.get_fields(),
+                'method': self.get_methods(),
+                'param': self.get_params()
+            },
+            'rename': {
+                'class': self.get_classes(),
+                'field': self.get_fields(),
+                'method': self.get_methods(),
+                'param': self.get_params()
+            },
+            'list': {
+                'class': self.get_classes(),
+                'classes': None,
+                'relation': self.get_classes(),
+                'relations': None
+            },
+            'help': {
+                'class': None,
+                'attribute': None,
+                'relation': None
+            },
+            'save': None,
+            'load': None,
+            'undo': None,
+            'redo': None,
+            'quit': None
+        })
+
+    def get_delete_relation(self) -> dict[str, set[str]]:
+        """
+        Get a dictionary of relations to be deleted between classes.
+
+        Returns:
+            dict[str, set[str]]: A dictionary where keys are class names (str) and values are sets
+                                containing names of classes (str) with existing relations to be deleted.
+        """
+        res = dict()
+        for uml_class in self.get_classes():
+            existing_relations = set()
+            for rel in self._diagram.get_all_relations():
+                if rel.get_src_name() == uml_class:
+                    existing_relations.add(rel.get_dst_name())
+            if existing_relations:
+                res[uml_class] = existing_relations
+        return res
+
+    def get_add_relation(self) -> dict[str, dict[str, str]]:
```

Comment:
separate out relation types (like you did before) so that you don't need to duplicate them for every possible relation. That will be horrible for performance on larger diagrams, and since all relation types will always be valid for a valid add there's no reason to attach them to each one individually.

Reply:
Not entirely sure if I'm understanding you, but I need to add the relations to the end of each last class for it to be autosuggested.